Fetch, save, and upload cluster trace logs from multiple services

chxchx opened the pull request:
@sebastienvas @yutongz 

This PR addresses the issue: [Enable cloud tracing in e2e framework #241](https://github.com/istio/istio/issues/241)

It fetches logs using Stackdriver golang client v2 and writes to log path just like test logs.
Other nit includes renaming some private functions in `istio/tests/e2e/framework/framework.go`. One might expect get*() only to access/retrieve members without modification. Renaming to pop*() and remove*() makes our intention unambiguous.

yutongz commented:
You may want to run bin/linters.sh and fix warnings.

sebastienvas commented on tests/e2e/framework/framework.go:
```diff
@@ -103,7 +103,7 @@ func (t *testCleanup) RegisterCleanable(c Cleanable) {
 }
 
-func (t *testCleanup) getCleanable() Cleanable {
+func (t *testCleanup) removeCleanable() Cleanable {
```
I find removeCleanble more confusing. I think pop would probably be best.

sebastienvas commented on tests/e2e.sh:
```diff
@@ -30,14 +30,18 @@ TESTS_TARGETS=($(bazel query 'tests(//tests/e2e/tests/...)'))
 
-
-
+# query project-id and put on environment
```
Let s not make this a default. Let s have the user set it only when testing on GCP.

sebastienvas replied:
Do not set the project id here. Have someone set it.

sebastienvas commented on tests/e2e/framework/testInfo.go:
```diff
@@ -111,6 +114,70 @@ func (t testInfo) Update(r int) error {
+func (t testInfo) FetchAndSaveClusterLogs() error {
+	ctx := context.Background()
+	glog.Info("Fetching cluster logs")
+	var PROJECT_ID string = *projectId
```
Why do you need to redeclare projectID ?

sebastienvas replied:
Also please use  MixedCaps or mixedCaps.

sebastienvas commented on tests/e2e/framework/testInfo.go:
```diff
@@ -111,6 +114,70 @@ func (t testInfo) Update(r int) error {
+		defer f.Close()
+		// fetch log entries with pagination 
+		var entries []*logging.Entry
+		const PAGE_SIZE = 50
```
You probably define this const at the beginning of the file.

chxchx replied:
Done.

sebastienvas commented on tests/e2e/framework/testInfo.go:
```diff
@@ -111,6 +114,70 @@ func (t testInfo) Update(r int) error {
+	// This list of log identifiers is cherry-picked based on Dev team's interest
+	// To access the comprehensive list, try in your shell
+	// 		$ gcloud beta logging logs list
+	LOG_IDS := []string{
```
You might want to define those as a const at the beginning of the file and add a flag for this.

chxchx replied:
I moved this to the front in latest commit. Do you mind clarifying what it means by "add a flag" for this?

sebastienvas commented:
Can you also add information about deployments, ie list all pods, services, ingress, etc uses in test and save the yaml files somewhere to help debugging ?

chxchx commented:
I add another "log_provider" flag as @rshriram suggested and the features in this PR will not be triggered unless "--log_provider=stackdriver" is specified.

Also improved the util.Shell() function to not just run the command but do so in a shell so that one may take advantage of the shell script features such as pipe(|), dup(<, >), fork(&), etc. This makes fetching deployment information (pods, services, ingress, etc, per request by @sebastienvas ) much simpler.

rshriram commented on tests/e2e/framework/testInfo.go:
```diff
@@ -28,18 +28,41 @@ import (
+		"istio-ingress",
+		"mixer",
+		"prometheus",
+		"statesd-to-prometheus",
```
Typo

rshriram commented on tests/e2e/framework/testInfo.go:
```diff
@@ -28,18 +28,41 @@ import (
 const (
 	tmpPrefix   = "istio.e2e."
 	idMaxLength = 36
+	pageSize = 20
```
Please add comment describing what is pageSize

sebastienvas commented on tests/e2e/framework/testInfo.go:
```diff
@@ -28,18 +28,41 @@ import (
 
 var (
 	logsBucketPath = flag.String("logs_bucket_path", "", "Cloud Storage Bucket path to use to store logs")
+	projectID = flag.String("project_id", "istio-testing", "Project ID")
```
Please set this to "". This value only make sense for Google.

sebastienvas commented on tests/e2e/framework/testInfo.go:
```diff
@@ -111,6 +134,71 @@ func (t testInfo) Update(r int) error {
 	return t.createStatusFile(r)
 }
 
+func (t testInfo) FetchAndSaveClusterLogs() error {
```
Check if projectID is not empty before starting.

chxchx replied:
I reverted e2e.sh and extract to another shell script (getProjectID.sh) the section that gets project id with gcloud command. 
Here, I check if projectID is empty. If so, I call getProjectID. Users can always supersede this value with --project_id flag, but reading the project id is important because the user may have multiple clusters configured and the getProjectID script always identifies the right cluster where the tests are being run, so users do not have to constantly keep tract this value.

istio-testing commented:
Jenkins job istio/presubmit passed

sebastienvas commented on tests/e2e/framework/testInfo.go:
```diff
@@ -111,6 +134,87 @@ func (t testInfo) Update(r int) error {
+func (t testInfo) FetchAndSaveClusterLogs() error {
+	// initialize projectID if user did not specify
+	if *projectID == "" {
+		pjID, err := util.Shell("getProjectID.sh")
```
Sorry If I confused you. If projectID is not set, you should just skip this (ie not do anything). You don't need tests/e2e/framework/getProjectID.sh script either. You might want to set it in Jenkinsfile when you call e2e.sh in https://github.com/istio/istio/blob/master/Jenkinsfile#L51. You can use env.PROJECT inside Jenkinsfile to find the project.

istio-testing commented:
Jenkins job istio/presubmit passed

istio-testing commented:
Jenkins job istio/presubmit passed

chxchx commented:
getProjectID.sh is removed and jenkinsfile updated. I tried env.PROJECT but it didn't work. Assuming jenkins only runs on istio-testing, I hard coded the string. 

Much time was invested to get log ids at runtime thru stackdriver api. Turns out one need v2 api to do so, which is incompatible with what I have been using. Migration has been painful but is done.

sebastienvas commented on Jenkinsfile:
```diff
@@ -54,14 +54,18 @@ def smokeTest(gitUtils, bazel, utils) {
-              "--istioctl_url=${istioctlUrl}"
+              "--istioctl_url=${istioctlUrl} " +
+              "--log_provider=${logHost} " +
+              "--project_id=${projID}"
```
Do you have a run when env.PROJECT did not work. There is no reason why it should not.

istio-testing commented:
@chxchx: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 689e7ceabb2bec6a077d069fcc5c9b91f91e373a | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/440/istio-presubmit/6/) | `@istio-testing bazel test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
Jenkins job istio/presubmit passed

istio-testing commented:
Jenkins job istio/presubmit passed

chxchx commented on Jenkinsfile:
```diff
@@ -44,7 +44,9 @@ def smokeTest(gitUtils, bazel, utils) {
-    def e2eArgs = "--logs_bucket_path ${gitUtils.logsPath()} "
+    def logHost = 'stackdriver'
+    def projID = 'istio-testing'
+    def e2eArgs = "--logs_bucket_path ${gitUtils.logsPath()} --log_provider=${logHost} --project_id=${projID} "
```
@sebastienvas Is this the right way to set flag value? Even when I hardcoded string like this, the latest Jenkins run ([console output](https://testing.istio.io/job/istio/job/presubmit/777/console)) still seems to fail to pick these value up, since I print these values out in framework.go. Any suspection on what could be wrong?

yutongz replied:
Take a look at line 23, your change only affects SMOKE_TEST for other repos. While here, we are running PRESUBMIT at line 20.

istio-testing commented:
Jenkins job istio/presubmit passed

istio-testing commented:
Jenkins job istio/presubmit passed

istio-testing commented:
Jenkins job istio/presubmit passed

istio-testing commented:
Jenkins job istio/presubmit passed

istio-testing commented:
Jenkins job istio/presubmit passed

istio-testing commented:
Jenkins job istio/presubmit passed

yutongz commented on tests/e2e/framework/testInfo.go:
```diff
@@ -51,7 +51,7 @@ var (
 	tmpPrefix   = "istio.e2e."
 	idMaxLength = 36
-	pageSize    = 100 // number of log entries for each paginated request to fetch logs
+	pageSize    = 500 // number of log entries for each paginated request to fetch logs
```
Just curiosity, why keep extending pageSize, do you think it's necessary to make it configurable?

chxchx replied:
As we discussed earlier, the issue here is that the log api endpoint rejects our request if requests are sent too frequently. Yutong and I agree that we can try a really large number.

yutongz commented on tests/e2e/framework/testInfo.go:
```diff
@@ -111,6 +123,102 @@ func (t testInfo) Update(r int) error {
+	for _, resrc := range resources {
+		glog.Info(fmt.Sprintf("Fetching deployment info on %s\n", resrc))
+		path := filepath.Join(t.LogsPath, fmt.Sprintf("%s.yaml", resrc))
+		if yaml, err0 := util.Shell(fmt.Sprintf("kubectl get %s -o yaml", resrc)); err0 == nil {
```
handel err != nil first

yutongz replied:
And do you need to specify namespace?

chxchx replied:
If you take a look at line 141 to 144, the namespace is passed in from commend line and is used to filter/select log entries of that namespace only. 

Also inverted the if branches so that conditions always check for err != nil. I agree that doing so improves readability.

istio-testing commented:
Jenkins job istio/presubmit passed

istio-testing commented:
Jenkins job istio/presubmit passed

chxchx commented:
@yutongz @sebastienvas Please take another look. Thanks!

sebastienvas commented on tests/e2e/framework/testInfo.go:
```diff
@@ -111,6 +123,102 @@ func (t testInfo) Update(r int) error {
+		glog.Info(fmt.Sprintf("Fetching deployment info on %s\n", resrc))
+		path := filepath.Join(t.LogsPath, fmt.Sprintf("%s.yaml", resrc))
+		if yaml, err0 := util.Shell(fmt.Sprintf("kubectl get %s -n %s -o yaml", resrc, *namespace)); err0 != nil {
+			errMsg += err0.Error() + "\n"
```
Instead of appending, please use multierror like so https://github.com/istio/istio/blob/master/tests/e2e/apps/hop/app.go#L32

sebastienvas commented:
I just submitted https://github.com/istio/istio/pull/464 to enable Prow. Could you enable your flags in https://github.com/istio/istio/blob/master/prow/istio-presubmit.sh#L40.

yutongz commented on tests/e2e/framework/testInfo.go:
```diff
@@ -111,6 +123,102 @@ func (t testInfo) Update(r int) error {
+		filter := fmt.Sprintf(
+			`logName = "%s" AND
+			resource.labels.namespace_id = "%s"`,
+			logName, *namespace)
```
I think the better way to get namespace is to make it as a member of testInfo

chxchx replied:
I too agree that dereferencing *namespace directly creates confusion on the code structure as it may be unclear where this value is initialized. However, I think passing namespace as a param to FetchAndSaveClusterLogs() is better than declaring as member testInfo. For one, namespace is already a member (read from command-line flag) of KubeInfo, and calling newKubeInfo requires a testInfo, so defining namespace in testInfo creates a cyclic dependency. Also, logically, namespace is a property of kubectl and less so for testInfo.

yutongz commented on tests/e2e/framework/testInfo.go:
```diff
@@ -111,6 +123,102 @@ func (t testInfo) Update(r int) error {
+	for _, resrc := range resources {
+		glog.Info(fmt.Sprintf("Fetching deployment info on %s\n", resrc))
+		path := filepath.Join(t.LogsPath, fmt.Sprintf("%s.yaml", resrc))
+		if yaml, err0 := util.Shell(fmt.Sprintf("kubectl get %s -n %s -o yaml", resrc, *namespace)); err0 != nil {
```
Same here, make namespace a member of testInfo

istio-testing commented:
@chxchx: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 7e5992faed83ce15781b725461a3daa94397f455 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/440/istio-presubmit/35/) | `@istio-testing bazel test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
Jenkins job istio/presubmit passed

istio-testing commented:
Jenkins job istio/presubmit passed

istio-testing commented:
@chxchx: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | db836d894ae66871ff02223f8a378ca7552986c6 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/440/istio-presubmit/40/) | `@istio-testing bazel test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@chxchx: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | fe1592de9c544e2777edbe6f74629d9531633144 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/440/istio-presubmit/41/) | `@istio-testing bazel test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@chxchx: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 5bf887c667f037559294bfdf7f2a90da884ad38d | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/440/istio-presubmit/42/) | `@istio-testing bazel test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
@chxchx: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 0765a65a19f9aa8117b875e8fcbf3f4e5aa30927 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/440/istio-presubmit/43/) | `@istio-testing bazel test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
Jenkins job istio/presubmit passed

istio-testing commented:
Jenkins job istio/presubmit passed

istio-testing commented:
Jenkins job istio/presubmit passed

chxchx commented:
@sebastienvas @yutongz I have addressed your comments in the most part, except that I had two problems with Prow.
First, I had trouble getting project id. Back in Jenkins we use env.PROJECT and I was wondering if we have similar setup in Prow. 
Second, using a hardcoded projectid, I was able to pass the prow presubmit but the steps to fetch log from stackdriver returns ` Error during log saving: rpc error: code = PermissionDenied desc = The caller does not have permission`, which is confusing since jenkins has never had this problem before.

istio-testing commented:
Jenkins job istio/presubmit passed

sebastienvas commented on tests/e2e/framework/testInfo.go:
```diff
@@ -121,6 +134,102 @@ func (t testInfo) Update(r int) error {
+				}
+			}
+		}
+	}
```
skip the if section and:
return multiErr

sebastienvas commented on tests/e2e/framework/testInfo.go:
```diff
@@ -121,6 +134,102 @@ func (t testInfo) Update(r int) error {
+		Parent: "projects/" + *projectID,
+	}
+	it := loggingClient.ListLogs(ctx, req)
+	multiErr := fmt.Errorf("")
```
var multiErr error

sebastienvas commented:
I think the right way to set the project would be to do it in Prow. Let's ask about this to @nlandolfi on Monday. For your other question, the user running Prow jobs, must not have permission on the Project. I think the right way to do this would be to create a role for run e2e test and give it what roles are necessary. Let me show you this on Monday.

istio-testing commented:
Jenkins job istio/presubmit passed

nlandolfi commented on tests/e2e/framework/testInfo.go:
```diff
@@ -121,6 +134,99 @@ func (t testInfo) Update(r int) error {
+			return err
+		}
+		if i := strings.Index(logName, "presubmit"); i == -1 {
+			if err := fetchAndWrite(logName); err != nil {
```
might we be able to fetch logs concurrently by wrapping this logic in a lambda and using a waitgroup?

chxchx replied:
Thanks for the awesome idea! I just did exactly that and notice significant improvement in the completion time of e2e testing. Please take another look.

nlandolfi replied:
nice! big win

istio-testing commented:
Jenkins job istio/presubmit passed

istio-testing commented:
@chxchx: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | f3ccae4316cc76b9e57b64a437bc84df1b816d09 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/440/istio-presubmit/51/) | `@istio-testing bazel test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

sebastienvas commented:
@istio-testing test this

istio-testing commented:
Jenkins job istio/presubmit passed

istio-testing commented:
Jenkins job istio/presubmit passed

istio-testing commented:
@chxchx: The following test **failed**, say `/retest` to rerun them all:

Test name | Commit | Details | Rerun command
--- | --- | --- | ---
prow/istio-presubmit.sh | 3ccf0f7cf107c935a4785c63ddc27394bc38be27 | [link](https://k8s-gubernator.appspot.com/build/istio-prow/pull/istio_istio/440/istio-presubmit/53/) | `@istio-testing bazel test this`



<details>

Instructions for interacting with me using PR comments are available [here](https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://github.com/kubernetes/test-infra/blob/master/commands.md).
</details>
<!-- test report -->

istio-testing commented:
Jenkins job istio/presubmit passed

istio-testing commented:
Jenkins job istio/presubmit passed

istio-testing commented:
Jenkins job istio/presubmit passed

nlandolfi commented:
I may be misinterpreting, but although the Prow job passed, it appears the logs failed to upload https://storage.googleapis.com/istio-prow/pull/istio_istio/440/istio-presubmit/56/build-log.txt

grep for "Failed to save logs"

Perhaps this should return a non-zero exit status? Or should a failure here not matter?

sebastienvas commented:
I don t think we want to fail the test if we cannot get the logs. On success failing to get the logs should not block a PR, it matters in failure.

sebastienvas commented:
on your last commit it takes 20 minute per test to get the logs. I don t think that s acceptable. This tripples the presubmit time. I think we might want to keep the kubectl get, but disable the fetching from stackdriver until we understand why this is happening.

istio-testing commented:
Jenkins job istio/presubmit passed

istio-testing commented:
Jenkins job istio/presubmit passed

chxchx commented:
PTAL
Logs fetching has been slow despite edging towards stackdriver api quota limit. To mitigate extended delay, logs are only fetched if any test fails. 
Errors during log fetching now surfaces as warnings and will by no means block an otherwise successful PR.
